mu4/notation/internal/masternotation.cpp

```diff
@@ -87,7 +87,124 @@ mu::Ret MasterNotation::load(const io::path& path, const INotationReaderPtr& rea
+        score->staves().insert(staffIndex + rstaff, staff);
+
+        score->insertPart(part, staffIndex);
+
+        int sidx = score->staffIdx(part);
+        int eidx = sidx + part->nstaves();
+        for (Measure* m = score->firstMeasure(); m; m = m->nextMeasure()) {
+            m->cmdAddStaves(sidx, eidx, true);
+        }
+
+        staffIndex += rstaff;
+    }
+
+    numberInstrumentNames(score);
+}
+
+void MasterNotation::initStaff(Staff* staff, const instruments::InstrumentTemplate& instrumentTemplate,
+                               const instruments::StaffType* staffType, int cidx)
+{
+    const StaffType* pst = staffType ? staffType : instrumentTemplate.staffTypePreset;
+    if (!pst) {
+        pst = StaffType::getDefaultPreset(instrumentTemplate.staffGroup);
+    }
+
+    StaffType* stt = staff->setStaffType(Fraction(0, 1), *pst);
+    if (cidx >= instruments::MAX_STAVES) {
+        stt->setSmall(false);
+    } else {
+        stt->setSmall(instrumentTemplate.smallStaff[cidx]);
+        staff->setBracketType(0, instrumentTemplate.bracket[cidx]);
+        staff->setBracketSpan(0, instrumentTemplate.bracketSpan[cidx]);
+        staff->setBarLineSpan(instrumentTemplate.barlineSpan[cidx]);
+    }
+    staff->setDefaultClefType(instrumentTemplate.clefs[cidx]);
+}
+
+Instrument MasterNotation::instrumentFromTemplate(const instruments::InstrumentTemplate& instrumentTemplate) const
+{
+    Instrument instrument;
+    instrument.setAmateurPitchRange(instrumentTemplate.aPitchRange.min, instrumentTemplate.aPitchRange.max);
+    instrument.setProfessionalPitchRange(instrumentTemplate.pPitchRange.min, instrumentTemplate.pPitchRange.max);
+    for (StaffName sn : instrumentTemplate.longNames) {
+        instrument.addLongName(StaffName(sn.name(), sn.pos()));
+    }
+    for (StaffName sn : instrumentTemplate.shortNames) {
+        instrument.addShortName(StaffName(sn.name(), sn.pos()));
+    }
+    instrument.setTrackName(instrumentTemplate.trackName);
+    instrument.setTranspose(instrumentTemplate.transpose);
+    instrument.setInstrumentId(instrumentTemplate.musicXMLId);
+    if (instrumentTemplate.useDrumset) {
+        instrument.setDrumset(instrumentTemplate.drumset ? instrumentTemplate.drumset : smDrumset);
+    }
+    for (int i = 0; i < instrumentTemplate.staves; ++i) {
+        instrument.setClefType(i, instrumentTemplate.clefs[i]);
+    }
+    //    TODO
+    // instrument.setMidiActions(instrumentTemplate.midiActions);
+    instrument.setArticulation(instrumentTemplate.midiArticulations);
+    for (const instruments::Channel& c : instrumentTemplate.channels) {
+        instrument.appendChannel(new instruments::Channel(c));
+    }
+    instrument.setStringData(instrumentTemplate.stringData);
+    instrument.setSingleNoteDynamics(instrumentTemplate.singleNoteDynamics);
+    return instrument;
+}
+
+void MasterNotation::numberInstrumentNames(Ms::MasterScore* score)
+{
+    std::vector<QString> names;
+    std::vector<QString> firsts;
```

Comment:
It's not very clear what is the purpose of this vector by name

Reply:
I wrote this code—I guess it's just being moved here? Could call it `firstInstancesOfInstruments` or something similar.

Reply:
Yeah, much better